### DISCUSSION TITLE
How to make folder structure work on Windows?

### DISCUSSION THREAD

**Discussion by cwienands1** [icloud-photos-downloader/icloud_photos_downloader#564] [General]
I hope this is the right place to ask this question. I am trying to set up the prebuilt binary _icloudpd-1.8.0-windows-amd64.exe_ on a Windows machine. I'd like to customize the folder structure to be more coarse-grained. I tried the `--folder-structure` pattern with these two variants...
```
icloudpd-1.8.0-windows-amd64.exe  [...other params...]  --folder-structure {:%Y/%m}
icloudpd-1.8.0-windows-amd64.exe  [...other params...]  --folder-structure "{:%Y/%m}"
icloudpd-1.8.0-windows-amd64.exe  [...other params...]  --folder-structure "%Y/%Y-%m"
icloudpd-1.8.0-windows-amd64.exe  [...other params...]  --folder-structure "%Y\%Y-%m"
```
...but in all cases the downloader created only a single subfolder called `m` or `Y-m` respectively and downloaded all pictures into it. Somehow it doesn't recognize the pattern correctly. I already searched the _Issues_ and _Discussions_ sections but surprisingly, no one else seems to have this problem. Any advice? Thanks!

**Comment by cwienands1** [comment]
I did find one _issue_ that reported problems with the `--folder-structure` parameter when using _fishshell_. The solution seems to have been to doublequote the parameter, which is what I tried , too, to no avail. But that issue seems to indicate that the value passed into `--folder-structure` might (need to) be platform-specific. `DEBUG` log level is turned on but the code that interprets the folder structure doesn't print out anything.

**Comment by AndreyNikiforov** [comment]
@cwienands1 can you try to run with `--only-print-filenames`? Does it print correct/expected paths?

Another dimension is to check another version to validate if it is a regressions. 1.7.3 was the first with exe, 1.8.1 is the latest available

**Comment by cwienands1** [comment]
OK, I tried the following:
- Ran with 1.7.3 as well as 1.8.1.
- Ran with `--only-print-filenames`

Unfortunately no luck with 1.7.3, the behavior is still the same. The filenames in the output look like this one:
```
C:\Users\awien\Pictures\iCloud April\Y-m\IMG_6749.HEIC
```
Full command line is this, just in case I'm doing something wrong.
```
icloudpd-1.7.3-windows-amd64.exe -d "C:\Users\awien\Pictures\iCloud April" -u my@email.com --cookie-directory "C:\Users\awien\Pictures\iCloud April\Cookies" --recent 100 --auto-delete --no-progress-bar --folder-structure "{:%Y/%Y-%m}" --only-print-filenames
```

Time permitting I may install Python on this computer and try some debugging (no Python developer here, though).

**Reply by AndreyNikiforov** [reply]
@cwienands1 Just ran on my Windows 10 PC with latest with newly created folder structure matching yours:

```
d:\downloads\icloudpd-1.8.1-windows-amd64.exe --directory "d:\photos\test with space" --cookie-directory "d:\photos\test with space\Cookies" --username redacted@email.address  --folder-structure "{:%Y/%Y-%m}" --auto-delete --no-progress-bar --only-print-filenames --recent 10
iCloud Password:
d:\photos\test with space\2023\2023-02\21b5c002-8a87-4005-a009-cd6976c43379.mp4
d:\photos\test with space\2023\2023-01\69672177861__9FEAB77B-1740-4529-98FE-4573B3888EFE.HEIC
```

I am puzzled what else can be different...

**Comment by cwienands1** [comment]
Come to think of it, this my wife’s new Win 11 machine. I’ll try on my own Win 10 machine tomorrow and compare.

**Comment by cwienands1** [comment]
Mystery solved, and I can claim today's dummy award ;-) I had constructed the command line in a batch file, which I intend to use for scheduled syncs. It finally occurred that `%`s are special characters in batch files and need to be escaped with `%%`. Sorry for the trouble. But hopefully the next person making the same mistake will find this thread.

The value that I am using now in the batch file is: `--folder-structure "{:%%Y\%%Y-%%m}"`

**Reply by AndreyNikiforov** [reply]
> Mystery solved, and I can claim today's dummy award ;-) I had constructed the command line in a batch file, which I intend to use for scheduled syncs. It finally occurred that `%`s are special characters in batch files and need to be escaped with `%%`. Sorry for the trouble. But hopefully the next person making the same mistake will find this thread.
> 
> The value that I am using now in the batch file is: `--folder-structure "{:%%Y\%%Y-%%m}"`

Awesome!

What scheduling are you considering (goal, features, and implementation choice)? I am wondering because I thought about including scheduling into the app for some time #568

**Reply by cwienands1** [reply]
In the end this will be running on a Windows file server and the goal is that the pictures from our two iPhones magically show up in a network folder so we can periodically sort them into another folder structure “by occasion”. Today I’m using the iCloud client to do that but you can only link it to one account. I need to sync from two accounts simultaneously, which is where IPD comes into play. 

For scheduling I’m just thinking of running this once an hour, or every few hours. Not sure if there is some limitation on iCloud side.

**Reply by cwienands1** [reply]
Forgot to mention this… I am planning to incorporate the iCloud photo download into some larger automation that sorts and moves pictures, detects duplicates between our phones (all those pics we share via SMS and Whatsapp and end up on both phones) and such. At least that’s the plan. The reasons why I am mentioning this is because a scheduling feature that is built into IPD probably wouldn’t really help here. I’ll still need to create a scheduled job that combines photo download with all the other post processing. Just some input for your prioritization of work…

**Reply by AndreyNikiforov** [reply]
> Forgot to mention this… I am planning to incorporate the iCloud photo download into some larger automation that sorts and moves pictures, detects duplicates between our phones (all those pics we share via SMS and Whatsapp and end up on both phones) and such. At least that’s the plan. The reasons why I am mentioning this is because a scheduling feature that is built into IPD probably wouldn’t really help here. I’ll still need to create a scheduled job that combines photo download with all the other post processing. Just some input for your prioritization of work…

Thanks for sharing. I am curious of your plans and ideas from perspectives of the icloudpd project as well as managing my own collection of digital assets.

From the latter perspective, it seems to me that %% of dup photos is small in my case, so not worth extra deduping for the storage sake, while most photo management apps have deduping built in. The use case that is suboptimal in my case is selection of photos for printing through Shutterfly -- too many manual error-prone steps (beyond selection itself).
